### PR TITLE
[6.0] Add subquery support for from() and table() - OPTION 2

### PR DESCRIPTION
**Pull request by reinink**
This PR adds the ability to use subqueries with the database `table()` method and query builder `from()` method. This is an alternative approach to my other PR: #29594.

```php
$query->select('name')
    ->from(function ($query) {
        $query->something();
    }, 'users')
    ->get();

DB::table(function ($query) {
    $query->something();
}, 'users')->get();
```

What's nice about this approach is that you can alias regular table names as well:

```php
DB::table('users', 'alias')
```

Pinging @staudenmeir to verify that my approach to aliasing here is safe. Based on my tests, the aliased table name appears to be properly escaped.

As for breaking changes, I think the only one is the argument change on the `Illuminate/Database/Capsule/Manager@table` method.

### CODE REVIEW

File: src/Illuminate/Database/Capsule/Manager.php [hunk 1]
```diff
@@ -81,7 +81,7 @@ public static function connection($connection = null)
      * @param  string|null  $connection
      * @return \Illuminate\Database\Query\Builder
      */
-    public static function table($table, $connection = null)
+    public static function table($table, $as = null, $connection = null)
```
Comment (georgehanson) [inline src/Illuminate/Database/Capsule/Manager.php]: What is the `$as` parameter here for? I can't see it being used on the static method.

Comment (reinink) [reply src/Illuminate/Database/Capsule/Manager.php]: Why not? If this is the entry point into the query, where you specify the table, you may want to specify a subquery.

Comment (georgehanson) [reply src/Illuminate/Database/Capsule/Manager.php]: Sorry, I meant that the method definition is still

```php
return static::$instance->connection($connection)->table($table);
```

Should it be passing it like this instead?

```php
return static::$instance->connection($connection)->table($table, $as);
```

Comment (reinink) [reply src/Illuminate/Database/Capsule/Manager.php]: 🤦‍♂ 

Fixing.

Comment (reinink) [reply src/Illuminate/Database/Capsule/Manager.php]: Fixed. Thanks @georgehanson!

Comment (georgehanson) [reply src/Illuminate/Database/Capsule/Manager.php]: No problem, I'm not sure why the tests didn't catch that error though...

Might want to update the docblock too :)

Comment (reinink) [reply src/Illuminate/Database/Capsule/Manager.php]: > No problem, I'm not sure why the tests didn't catch that error though...

😬 

> Might want to update the docblock too :)

Ugh, I need more coffee. Thanks again, will do.

**Comment by staudenmeir** [comment]
When I suggested this approach, I didn't realize that `Manager` also has a `table()` method. I think this is a particularly nasty breaking change, as it only affects people who are using the `illuminate/database` package outside of Laravel. I imagine that they don't typically read the Laravel upgrade guide.

**Comment by reinink** [comment]
> as it only affects people who are using the illuminate/database package outside of Laravel.

Yup. That said, it seems kind of insane to cater to this though. 🤷‍♂

**Comment by mfn** [comment]
> I imagine that they don't typically read the Laravel upgrade guide.

They have to, there is no alternative.

I'm one of those users using `illuminate/database` outside Laravel.

The individual packages don't have proper release management (like individual changelog) so the only thing is the Laravel release guide itself.

**Comment by reinink** [comment]
Hmm, reflecting on this, I don't think this needs to be a breaking change. We can simply do a type check here to support the two argument version

```php
public static function table($table, $as = null, $connection = null) {

    if (is_string($as) && is_null($connection) {
        $connection = $as;
        $as = null;
    }

    // ...
}
```

I'll update the PR later.

**Comment by staudenmeir** [comment]
> Hmm, reflecting on this, I don't think this needs to be a breaking change. We can simply do a type check here to support the two argument version

Isn't `$as` always a string?

**Comment by driesvints** [comment]
@reinink it's always important that we also consider the components individually. One of the nice things about the framework is that you can re-use the components individually in custom build apps outside the framework. As long as we always add everything to the upgrade guide we should be good 👍

**Pull request merged**